### PR TITLE
Create USE Dashboard at startup

### PR DESCRIPTION
**Pull request by HeinrichHartmann**

### CODE REVIEW

**Comment by HeinrichHartmann** [comment]
Not read to merge, yet. Just sharing progress.

**Comment by HeinrichHartmann** [comment]
Latest PR introduced some merge conflicts which need work to resolve.
Also I am not sure if the new VM graph will work as is on all platforms.

File: src/content/files/cosi-install.sh [hunk 1]
```diff
@@ -939,13 +939,18 @@ cosi_register() {
         return
     fi
 
-    log "Running COSI registration script"
-
     [[ -x "$cosi_script" ]] || fail "Unable to find cosi command '${cosi_script}'"
 
+    echo
```
Comment (maier) [inline src/content/files/cosi-install.sh:995]: what log are these intended for? echo will only go to console, not a log.

File: src/util/lib/cosi/registration/graphs/index.js [hunk 2]
```diff
@@ -132,11 +132,7 @@ class Graphs extends Registration {
 
                 console.log(`\tFound ${templateType}-${templateId} ${template.file}`);
 
-                if ({}.hasOwnProperty.call(self.metrics, templateId)) {
-                    self.templates.push(templateId);
-                } else {
-                    console.log(`\t${chalk.yellow('Skipping')} ${templateType}-${templateId}, no metrics found for '${templateId}'.`);
-                }
+                self.templates.push(templateId);
```
Comment (maier) [inline src/util/lib/cosi/registration/graphs/index.js]: cosi should only create graphs where the metrics are available.

File: src/content/templates/graph-use_cpu.json [hunk 3]
```diff
@@ -0,0 +1,104 @@
+{
```
Comment (maier) [inline src/content/templates/graph-use_cpu.json]: this is graph-cpu the metrics come from the cpu plugin. if these are new graphs, add them to the existing graph-cpu template.

File: src/content/templates/graph-use_vm.json [hunk 4]
```diff
@@ -0,0 +1,148 @@
+{
```
Comment (maier) [inline src/content/templates/graph-use_vm.json]: this is graph-vm the metrics come from the vm plugin. if these are new graphs add them to the existing graph-vm template.

File: src/util/lib/cosi/registration/dashboards/index.js [hunk 5]
```diff
@@ -223,7 +223,7 @@ class Dashboards extends Registration {
         console.log(chalk.blue(this.marker));
         console.log(`Configuring dasbhoard`);
 
-        const templateMatch = template.file.match(/^template-dashboard-([^-]+)-(.+)\.json$/);
```
Comment (maier) [inline src/util/lib/cosi/registration/dashboards/index.js]: why is the dash being removed? the format for template files is template-<type>-<id>

File: src/util/bin/cosi-graph-update.js [hunk 6]
```diff
@@ -17,6 +17,7 @@ const Graph = require(path.join(cosi.lib_dir, 'graph'));
 app.
     version(cosi.app_version).
     usage('[options] <config_file>').
+    description('Update Circonus graph from local graph config file (/opt/circonus/cosi/registration/registration-*.json)').
```
Comment (maier) [inline src/util/bin/cosi-graph-update.js]: i wouldn't use the registration directory in the example. just something like `graph_configuration.json` remember, the registration directory is _documentation_ on what happened during the registration process. we shouldn't encourage doing things _in_ that directory. making a copy of a file, editing it, and using an update command is ok though.

File: src/util/lib/cosi/registration/graphs/index.js [hunk 7]
```diff
@@ -404,8 +404,11 @@ class Graphs extends Registration {
         assert.equal(typeof graphIdx, 'number', 'graphIdx is required');
 
         // flat list of full metric names metric_group`metric_name (fs`/sys/fs/cgroup`df_used_percent)
-        const metrics = Object.keys(this.metrics[template.id]).map((val) => {
-            return `${template.id}\`${val}`;
+        var metrics = [];
```
Comment (maier) [inline src/util/lib/cosi/registration/graphs/index.js]: don't use `var`

File: src/util/lib/cosi/registration/graphs/index.js [hunk 8]
```diff
@@ -404,8 +404,11 @@ class Graphs extends Registration {
         assert.equal(typeof graphIdx, 'number', 'graphIdx is required');
 
         // flat list of full metric names metric_group`metric_name (fs`/sys/fs/cgroup`df_used_percent)
-        const metrics = Object.keys(this.metrics[template.id]).map((val) => {
```
Comment (maier) [inline src/util/lib/cosi/registration/graphs/index.js]: what exactly was the problem?

File: src/content/templates/graph-network_summary.json [hunk 9]
```diff
@@ -0,0 +1,222 @@
+{
```
Comment (maier) [inline src/content/templates/graph-network_summary.json:1]: this is graph-if

File: src/util/lib/cosi/template/fetch/index.js [hunk 10]
```diff
@@ -156,7 +156,11 @@ class Fetch extends Events {
             }
 
             // list of all templates applicable to this host
-            const wantTemplates = [ 'check-system', 'worksheet-system' ];
```
Comment (maier) [inline src/util/lib/cosi/template/fetch/index.js]: just dashboard-system the others should be named appropriately after the plugins the metrics come from. if you want a load graph by default add a load plugin in NAD that is installed by default.

**Comment by maier** [comment]
if this concept only applicable to certain systems the templates can be put into directories specific to the dist, dist/ver, dist/ver/arch. we can add system type as well (e.g. Linux) if they are applicable to *all* derivatives of that type.

File: src/util/lib/cosi/registration/dashboards/index.js (hunk 5 above)
Comment (maier) [reply src/util/lib/cosi/registration/dashboards/index.js]: ah, i get it. for the new dashboard-system you created. make sense.

File: src/content/templates/graph-disk_summary.json [hunk 11]
```diff
@@ -0,0 +1,79 @@
+{
```
Comment (maier) [inline src/content/templates/graph-disk_summary.json]: graph-diskstats

File: src/content/files/cosi-install.sh (hunk 1 above)
Comment (HeinrichHartmann) [reply src/content/files/cosi-install.sh:995]: This helps to identify messages that are emitted by cosi-install.sh as opposed to (/setup/index.js, etc.) in the console.

The use of echo is precedented a few lines above:
https://github.com/HeinrichHartmann/circonus-one-step-install/blob/0089e9bfdbd9a17ff2446d4d6a55bae8ab1d0093/src/content/files/cosi-install.sh#L931

File: src/util/lib/cosi/registration/graphs/index.js (hunk 7 above)
Comment (HeinrichHartmann) [reply src/util/lib/cosi/registration/graphs/index.js]: The metrics reference is mutated on line 411. `const` won't allow this.

File: src/util/bin/cosi-graph-update.js (hunk 6 above)
Comment (HeinrichHartmann) [reply src/util/bin/cosi-graph-update.js]: Right. I just wanted to point the user at an example of a valid config file, which they could use to get started. Maybe `config-*` is better?

**Comment by HeinrichHartmann** [comment]
Two remarks that apply to multiple snippets:

(A) The USE Dashboard contains graphs that take data from multiple plugins. At the moment this is the `CPU Saturation` graph with metrics from `loadaverage.sh` and `cpu.sh`. I had to extend the graph templating to get this working. Even if we could avoid the issue here, by moving load averages to the CPU plugin, this will be impossible once we start exploiting eBPF metrics more (e.g. scheuler latency, network errors), etc. Graphs are not always dedicated to single plugins.

(B) We need to extend the "variable_metrics" feature for the Network and Disk graphs. We have support for creating different graphs for each disk/netowrk interface (let's call that "variable graphs"), but not to include a variable number of datapoints in a single graph ("variable datapoints"). The way this is realized in the PR is, that the datapoint JSON object itself gets a "variable" and "filter" properties similar to what the full graph template had. The pattern syntax for metric names stays the same. At the moment variable graphs do not mix with variable datapoints, so we need two different templates for disks and network interfaces. (This can be changed)

File: src/util/lib/cosi/registration/graphs/index.js (hunk 2 above)
Comment (HeinrichHartmann) [reply src/util/lib/cosi/registration/graphs/index.js]: OK this needs fixing. It was added to address (A) form below.

The exiting if-clause checks if we have a plugin/"metric group" that is named after the `templateId`, this is different from checking if there are metrics for the template. There might be very well, just not in that metric group.

I might be able to work around this by putting everyhing into the existing templates which have appropriate templateIds set...

File: src/content/templates/graph-use_cpu.json (hunk 3 above)
Comment (HeinrichHartmann) [reply src/content/templates/graph-use_cpu.json]: Done later in this PR.

File: src/content/templates/graph-use_vm.json (hunk 4 above)
Comment (HeinrichHartmann) [reply src/content/templates/graph-use_vm.json]: Done later in this PR.

File: src/util/lib/cosi/registration/dashboards/index.js (hunk 5 above)
Comment (HeinrichHartmann) [reply src/util/lib/cosi/registration/dashboards/index.js]: Exactly. We only have a single instance of the system dashboard.

File: src/content/templates/graph-network_summary.json (hunk 9 above)
Comment (HeinrichHartmann) [reply src/content/templates/graph-network_summary.json:1]: See (B). We can still merge it into graph-if, by extending the variable graph feature a little more.

File: src/util/lib/cosi/template/fetch/index.js (hunk 10 above)
Comment (HeinrichHartmann) [reply src/util/lib/cosi/template/fetch/index.js]: OK

**Comment by HeinrichHartmann** [comment]
Pushed revisions that address the review comments.

Ad A: Moved all graph into texisting templates: graph-if, graph-diskstats, etc.
Ad B: Extended the variable graph handling, to accomodate both use cases.

File: src/content/templates/graph-network_summary.json (hunk 9 above)
Comment (HeinrichHartmann) [reply src/content/templates/graph-network_summary.json:1]: merged into graph-if

File: src/util/lib/cosi/registration/graphs/index.js (hunk 8 above)
Comment (HeinrichHartmann) [reply src/util/lib/cosi/registration/graphs/index.js]: Visibility of metrics outside the metric group / plugin.

It turns out this is only needed for variable datapoints (not variable metrics/graphs), so I reverted to the old behavior here.

File: src/util/lib/cosi/registration/graphs/index.js (hunk 7 above)
Comment (HeinrichHartmann) [reply src/util/lib/cosi/registration/graphs/index.js]: Fixed.

File: src/util/lib/cosi/registration/dashboards/index.js (hunk 5 above)
Comment (HeinrichHartmann) [reply src/util/lib/cosi/registration/dashboards/index.js]: Made that regexp a little nicer.

**Comment by HeinrichHartmann** [comment]
Approved by @maier

**Pull request merged**